File: backend-api/app/services/benchmark_dataset_service.py
"""
Benchmark Dataset Service

This service manages research benchmark datasets for LLM safety evaluation,
including the Do-Not-Answer dataset from the paper "Do-Not-Answer: A Dataset
for Evaluating Safeguards in LLMs" (arxiv:2308.13387).

The service provides:
- Loading benchmark datasets from JSON files
- Filtering prompts by risk area, harm type, severity
- Generating test batches for jailbreak evaluation
- Integrating with the existing jailbreak database
"""

import json
import logging
import random
from dataclasses import dataclass, field
from enum import Enum
from pathlib import Path
from typing import Any

from pydantic import BaseModel

logger = logging.getLogger(__name__)

# Default data directory for benchmarks
BENCHMARKS_DIR = Path(__file__).parent.parent.parent.parent / "data" / "benchmarks"


class RiskArea(str, Enum):
    """Risk areas from the Do-Not-Answer taxonomy."""

    INFORMATION_HAZARDS = "I"
    MALICIOUS_USES = "II"
    DISCRIMINATION_TOXICITY = "III"
    MISINFORMATION = "IV"
    HUMAN_CHATBOT_INTERACTION = "V"


class Severity(str, Enum):
    """Severity levels for harmful prompts."""

    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


class BenchmarkPrompt(BaseModel):
    """A single prompt from a benchmark dataset."""

    id: str
    risk_area: str
    harm_type: str
    category: str
    subcategory: str
    prompt: str
    expected_behavior: str = "refuse"
    severity: str = "high"
    metadata: dict[str, Any] = {}


class BenchmarkDataset(BaseModel):
    """A complete benchmark dataset with metadata."""

    name: str
    version: str
    description: str = ""
    paper_title: str = ""
    paper_url: str = ""
    license: str = ""
    prompts: list[BenchmarkPrompt] = []
    taxonomy: dict[str, Any] = {}
    statistics: dict[str, Any] = {}


class HarmType(BaseModel):
    """A harm type from the taxonomy."""

    id: str
    name: str
    subcategories: list[str] = []
    example_prompts: list[str] = []


class RiskAreaInfo(BaseModel):
    """Information about a risk area."""

    id: str
    name: str
    description: str = ""
    harm_types: list[HarmType] = []


@dataclass
class BenchmarkDatasetService:
    """
    Service for managing benchmark datasets for LLM safety evaluation.

    Loads and manages datasets like Do-Not-Answer for testing jailbreak
    resistance and safety guardrails in language models.
    """

    benchmarks_dir: Path = field(default_factory=lambda: BENCHMARKS_DIR)
    _datasets: dict[str, BenchmarkDataset] = field(default_factory=dict)
    _loaded: bool = False

    def __post_init__(self) -> None:
        """Initialize the service and load available datasets."""
        self._datasets = {}
        self._loaded = False

    def load_datasets(self) -> None:
        """Load all available benchmark datasets from the benchmarks directory."""
        if self._loaded:
            return

        if not self.benchmarks_dir.exists():
            logger.warning("Benchmarks directory not found: %s", self.benchmarks_dir)
            self._loaded = True
            return

        # Load each dataset directory
        for dataset_dir in self.benchmarks_dir.iterdir():
            if dataset_dir.is_dir():
                self._load_dataset(dataset_dir)

        self._loaded = True
        logger.info("Loaded %d benchmark datasets", len(self._datasets))

    def _load_dataset(self, dataset_dir: Path) -> None:
        """Load a single benchmark dataset from a directory."""
        taxonomy_file = dataset_dir / "taxonomy.json"
        prompts_file = dataset_dir / "prompts.json"

        if not taxonomy_file.exists() and not prompts_file.exists():
            logger.warning("No taxonomy or prompts file found in %s", dataset_dir)
            return

        dataset_name = dataset_dir.name

        try:
            # Load taxonomy if available
            taxonomy = {}
            paper_info = {}
            statistics = {}

            if taxonomy_file.exists():
                with open(taxonomy_file, encoding="utf-8") as f:
                    taxonomy_data = json.load(f)
                    taxonomy = taxonomy_data.get("taxonomy", {})
                    paper_info = taxonomy_data.get("paper", {})
                    statistics = taxonomy_data.get("statistics", {})

            # Load prompts if available
            prompts = []
            if prompts_file.exists():
                with open(prompts_file, encoding="utf-8") as f:
                    prompts_data = json.load(f)
                    for p in prompts_data.get("prompts", []):
                        prompts.append(
                            BenchmarkPrompt(
                                id=p.get("id", ""),
                                risk_area=p.get("risk_area", ""),
                                harm_type=p.get("harm_type", ""),
                                category=p.get("category", ""),
                                subcategory=p.get("subcategory", ""),
                                prompt=p.get("prompt", ""),
                                expected_behavior=p.get("expected_behavior", "refuse"),
                                severity=p.get("severity", "high"),
                                metadata=p.get("metadata", {}),
                            )
                        )

            # Create dataset object
            dataset = BenchmarkDataset(
                name=taxonomy_data.get("name", dataset_name) if taxonomy_file.exists() else dataset_name,
                version=taxonomy_data.get("version", "1.0.0") if taxonomy_file.exists() else "1.0.0",
                description=taxonomy_data.get("description", "") if taxonomy_file.exists() else "",
                paper_title=paper_info.get("title", ""),
                paper_url=paper_info.get("url", ""),
                license=taxonomy_data.get("license", "") if taxonomy_file.exists() else "",
                prompts=prompts,
                taxonomy=taxonomy,
                statistics=statistics,
            )

            self._datasets[dataset_name] = dataset
            logger.info(
                "Loaded dataset '%s' with %d prompts", dataset_name, len(prompts)
            )

        except Exception as e:
            logger.error("Failed to load dataset from %s: %s", dataset_dir, e)

    def get_datasets(self) -> list[str]:
        """Get list of available dataset names."""
        self.load_datasets()
        return list(self._datasets.keys())

    def get_dataset(self, name: str) -> BenchmarkDataset | None:
        """Get a specific dataset by name."""
        self.load_datasets()
        return self._datasets.get(name)

    def get_prompts(
        self,
        dataset_name: str,
        risk_area: str | None = None,
        harm_type: str | None = None,
        severity: str | None = None,
        limit: int | None = None,
        shuffle: bool = False,
    ) -> list[BenchmarkPrompt]:
        """
        Get prompts from a dataset with optional filtering.

        Args:
            dataset_name: Name of the dataset to query
            risk_area: Filter by risk area (e.g., "I", "II", etc.)
            harm_type: Filter by harm type (e.g., "I-1", "II-2", etc.)
            severity: Filter by severity level
            limit: Maximum number of prompts to return
            shuffle: Whether to shuffle the results

        Returns:
            List of matching prompts
        """
        self.load_datasets()

        dataset = self._datasets.get(dataset_name)
        if not dataset:
            return []

        prompts = dataset.prompts

        # Apply filters
        if risk_area:
            prompts = [p for p in prompts if p.risk_area == risk_area]

        if harm_type:
            prompts = [p for p in prompts if p.harm_type == harm_type]

        if severity:
            prompts = [p for p in prompts if p.severity == severity]

        # Shuffle if requested
        if shuffle:
            prompts = list(prompts)  # Make a copy
            random.shuffle(prompts)

        # Apply limit
        if limit:
            prompts = prompts[:limit]

        return prompts

    def get_random_prompts(
        self,
        dataset_name: str,
        count: int = 10,
        risk_area: str | None = None,
        severity: str | None = None,
    ) -> list[BenchmarkPrompt]:
        """Get random prompts from a dataset."""
        return self.get_prompts(
            dataset_name=dataset_name,
            risk_area=risk_area,
            severity=severity,
            limit=count,
            shuffle=True,
        )

    def get_taxonomy(self, dataset_name: str) -> dict[str, Any]:
        """Get the taxonomy for a dataset."""
        self.load_datasets()

        dataset = self._datasets.get(dataset_name)
        if not dataset:
            return {}

        return dataset.taxonomy

    def get_risk_areas(self, dataset_name: str) -> list[RiskAreaInfo]:
        """Get all risk areas for a dataset."""
        taxonomy = self.get_taxonomy(dataset_name)
        if not taxonomy:
            return []

        risk_areas = []
        for ra in taxonomy.get("risk_areas", []):
            harm_types = []
            for ht in ra.get("harm_types", []):
                harm_types.append(
                    HarmType(
                        id=ht.get("id", ""),
                        name=ht.get("name", ""),
                        subcategories=ht.get("subcategories", []),
                        example_prompts=ht.get("example_prompts", []),
                    )
                )

            risk_areas.append(
                RiskAreaInfo(
                    id=ra.get("id", ""),
                    name=ra.get("name", ""),
                    description=ra.get("description", ""),
                    harm_types=harm_types,
                )
            )

        return risk_areas

    def get_statistics(self, dataset_name: str) -> dict[str, Any]:
        """Get statistics for a dataset."""
        self.load_datasets()

        dataset = self._datasets.get(dataset_name)
        if not dataset:
            return {}

        return dataset.statistics

    def generate_test_batch(
        self,
        dataset_name: str,
        batch_size: int = 10,
        balanced: bool = True,
        risk_areas: list[str] | None = None,
        severities: list[str] | None = None,
    ) -> list[BenchmarkPrompt]:
        """
        Generate a balanced test batch from a dataset.

        Args:
            dataset_name: Name of the dataset
            batch_size: Total number of prompts to include
            balanced: Whether to balance across risk areas
            risk_areas: Specific risk areas to include (None = all)
            severities: Specific severities to include (None = all)

        Returns:
            List of prompts for testing
        """
        self.load_datasets()

        dataset = self._datasets.get(dataset_name)
        if not dataset:
            return []

        prompts = dataset.prompts

        # Filter by risk areas if specified
        if risk_areas:
            prompts = [p for p in prompts if p.risk_area in risk_areas]

        # Filter by severities if specified
        if severities:
            prompts = [p for p in prompts if p.severity in severities]

        if not prompts:
            return []

        if balanced:
            # Get unique risk areas in filtered prompts
            unique_areas = list({p.risk_area for p in prompts})
            per_area = max(1, batch_size // len(unique_areas))

            batch = []
            for area in unique_areas:
                area_prompts = [p for p in prompts if p.risk_area == area]
                random.shuffle(area_prompts)
                batch.extend(area_prompts[:per_area])

            # Fill remaining slots randomly
            remaining = batch_size - len(batch)
            if remaining > 0:
                unused = [p for p in prompts if p not in batch]
                random.shuffle(unused)
                batch.extend(unused[:remaining])

            return batch[:batch_size]
        else:
            # Simple random selection
            prompts_copy = list(prompts)
            random.shuffle(prompts_copy)
            return prompts_copy[:batch_size]

    def export_prompts_for_jailbreak(
        self,
        dataset_name: str,
        risk_area: str | None = None,
        limit: int | None = None,
    ) -> list[dict[str, Any]]:
        """
        Export prompts in a format suitable for jailbreak testing.

        Returns prompts formatted for use with the AutoDAN and other
        jailbreak generation engines.
        """
        prompts = self.get_prompts(
            dataset_name=dataset_name,
            risk_area=risk_area,
            limit=limit,
            shuffle=True,
        )

        return [
            {
                "id": p.id,
                "request": p.prompt,
                "category": p.category,
                "risk_area": p.risk_area,
                "harm_type": p.harm_type,
                "expected_behavior": p.expected_behavior,
                "severity": p.severity,
                "metadata": {
                    "source": "benchmark",
                    "dataset": dataset_name,
                    "subcategory": p.subcategory,
                },
            }
            for p in prompts
        ]


# Global singleton instance
_benchmark_service: BenchmarkDatasetService | None = None


def get_benchmark_service() -> BenchmarkDatasetService:
    """Get the global benchmark dataset service instance."""
    global _benchmark_service
    if _benchmark_service is None:
        _benchmark_service = BenchmarkDatasetService()
    return _benchmark_service

File: frontend/src/lib/services/benchmark-dataset-service.ts
/**
 * Benchmark Dataset Service
 *
 * Service for interacting with the benchmark datasets API.
 * Provides access to Do-Not-Answer and other safety benchmark datasets
 * for adversarial testing and jailbreak evaluation.
 */

import { enhancedApi as api, apiClient } from "@/lib/api-enhanced";

// Environment check for debug logging
const isDevelopment = process.env.NODE_ENV === "development";

/**
 * Debug logger for Benchmark Dataset service.
 */
const debugLog = {
  log: (...args: unknown[]) => isDevelopment && console.log("[BenchmarkDataset]", ...args),
  error: (...args: unknown[]) => console.error("[BenchmarkDataset]", ...args),
  warn: (...args: unknown[]) => isDevelopment && console.warn("[BenchmarkDataset]", ...args),
};

/**
 * Risk area information from the Do-Not-Answer taxonomy
 */
export interface RiskAreaInfo {
  id: string;
  name: string;
  description: string;
  harm_types: HarmTypeInfo[];
  // Additional properties used by components
  display_name?: string;
  prompt_count?: number;
}

/**
 * Harm type information from the Do-Not-Answer taxonomy
 */
export interface HarmTypeInfo {
  id: string;
  name: string;
  description: string;
  example_behaviors: string[];
}

/**
 * A single benchmark prompt
 */
export interface BenchmarkPrompt {
  id: string;
  risk_area: string;
  harm_type: string;
  category: string;
  subcategory?: string;
  prompt: string;
  expected_behavior: string;
  severity: "low" | "medium" | "high" | "critical";
  metadata?: Record<string, unknown>;
}

/**
 * Summary of a benchmark dataset
 */
export interface BenchmarkDatasetSummary {
  name: string;
  display_name: string;
  description: string;
  total_prompts: number;
  risk_areas: number;
  harm_types: number;
  source: string;
  paper_url?: string;
  license?: string;
}

/**
 * Full benchmark dataset with all prompts and taxonomy
 */
export interface BenchmarkDataset {
  name: string;
  display_name: string;
  description: string;
  version: string;
  source: string;
  paper_url?: string;
  license?: string;
  prompts: BenchmarkPrompt[];
  taxonomy: {
    risk_areas: RiskAreaInfo[];
  };
  statistics: {
    total_prompts: number;
    prompts_by_risk_area: Record<string, number>;
    prompts_by_severity: Record<string, number>;
  };
}

/**
 * Statistics for a benchmark dataset
 */
export interface BenchmarkStatistics {
  total_prompts: number;
  prompts_by_risk_area: Record<string, number>;
  prompts_by_harm_type: Record<string, number>;
  prompts_by_severity: Record<string, number>;
  average_severity_score: number;
  // Additional properties used by components
  risk_area_counts?: Record<string, number>;
  harm_type_counts?: Record<string, number>;
  severity_counts?: Record<string, number>;
}

/**
 * Test batch configuration
 */
export interface TestBatchConfig {
  count: number;
  risk_areas?: string[];
  harm_types?: string[];
  min_severity?: "low" | "medium" | "high" | "critical";
  shuffle?: boolean;
}

/**
 * Test batch response
 */
export interface TestBatchResponse {
  dataset: string;
  prompts: BenchmarkPrompt[];
  count: number;
  filters_applied: {
    risk_areas?: string[];
    harm_types?: string[];
    min_severity?: string;
  };
}

/**
 * Export configuration for jailbreak testing
 */
export interface ExportConfig {
  format: "json" | "csv" | "jsonl";
  include_metadata?: boolean;
  risk_areas?: string[];
  harm_types?: string[];
}

/**
 * Benchmark Dataset Service class
 */
class BenchmarkDatasetService {
  private readonly baseUrl = "/benchmark-datasets";

  /**
   * List all available benchmark datasets
   */
  async listDatasets(): Promise<BenchmarkDatasetSummary[]> {
    const startTime = Date.now();
    const endpoint = this.baseUrl;

    debugLog.log("Fetching benchmark datasets list");

    try {
      const response = await api.get<{ datasets: BenchmarkDatasetSummary[] }>(endpoint);
      const duration = Date.now() - startTime;
      debugLog.log(`Fetched ${response.datasets?.length || 0} datasets in ${duration}ms`);
      return response.datasets || [];
    } catch (error) {
      debugLog.error("Failed to fetch datasets:", error);
      throw error;
    }
  }

  /**
   * Get details for a specific dataset
   */
  async getDataset(datasetName: string): Promise<BenchmarkDataset> {
    const startTime = Date.now();
    const endpoint = `${this.baseUrl}/${encodeURIComponent(datasetName)}`;

    debugLog.log(`Fetching dataset: ${datasetName}`);

    try {
      const response = await api.get<BenchmarkDataset>(endpoint);
      const duration = Date.now() - startTime;
      debugLog.log(`Fetched dataset ${datasetName} in ${duration}ms`);
      return response;
    } catch (error) {
      debugLog.error(`Failed to fetch dataset ${datasetName}:`, error);
      throw error;
    }
  }

  /**
   * Get prompts from a dataset with optional filtering
   */
  async getPrompts(
    datasetName: string,
    options?: {
      risk_area?: string;
      harm_type?: string;
      severity?: string;
      limit?: number;
      offset?: number;
    }
  ): Promise<{ prompts: BenchmarkPrompt[]; total: number }> {
    const startTime = Date.now();
    const params = new URLSearchParams();

    if (options?.risk_area) params.set("risk_area", options.risk_area);
    if (options?.harm_type) params.set("harm_type", options.harm_type);
    if (options?.severity) params.set("severity", options.severity);
    if (options?.limit) params.set("limit", options.limit.toString());
    if (options?.offset) params.set("offset", options.offset.toString());

    const endpoint = `${this.baseUrl}/${encodeURIComponent(datasetName)}/prompts?${params.toString()}`;

    debugLog.log(`Fetching prompts from ${datasetName}`, options);

    try {
      const response = await api.get<{ prompts: BenchmarkPrompt[]; total: number }>(endpoint);
      const duration = Date.now() - startTime;
      debugLog.log(`Fetched ${response.prompts?.length || 0} prompts in ${duration}ms`);
      return response;
    } catch (error) {
      debugLog.error(`Failed to fetch prompts from ${datasetName}:`, error);
      throw error;
    }
  }

  /**
   * Get random prompts from a dataset
   */
  async getRandomPrompts(
    datasetName: string,
    count: number = 5,
    options?: {
      risk_area?: string;
      harm_type?: string;
    }
  ): Promise<BenchmarkPrompt[]> {
    const startTime = Date.now();
    const params = new URLSearchParams();
    params.set("count", count.toString());

    if (options?.risk_area) params.set("risk_area", options.risk_area);
    if (options?.harm_type) params.set("harm_type", options.harm_type);

    const endpoint = `${this.baseUrl}/${encodeURIComponent(datasetName)}/random?${params.toString()}`;

    debugLog.log(`Fetching ${count} random prompts from ${datasetName}`);

    try {
      const response = await api.get<{ prompts: BenchmarkPrompt[] }>(endpoint);
      const duration = Date.now() - startTime;
      debugLog.log(`Fetched ${response.prompts?.length || 0} random prompts in ${duration}ms`);
      return response.prompts || [];
    } catch (error) {
      debugLog.error(`Failed to fetch random prompts from ${datasetName}:`, error);
      throw error;
    }
  }

  /**
   * Get the taxonomy for a dataset
   */
  async getTaxonomy(datasetName: string): Promise<{ risk_areas: RiskAreaInfo[] }> {
    const startTime = Date.now();
    const endpoint = `${this.baseUrl}/${encodeURIComponent(datasetName)}/taxonomy`;

    debugLog.log(`Fetching taxonomy for ${datasetName}`);

    try {
      const response = await api.get<{ risk_areas: RiskAreaInfo[] }>(endpoint);
      const duration = Date.now() - startTime;
      debugLog.log(`Fetched taxonomy in ${duration}ms`);
      return response;
    } catch (error) {
      debugLog.error(`Failed to fetch taxonomy for ${datasetName}:`, error);
      throw error;
    }
  }

  /**
   * Get risk areas for a dataset
   */
  async getRiskAreas(datasetName: string): Promise<RiskAreaInfo[]> {
    const startTime = Date.now();
    const endpoint = `${this.baseUrl}/${encodeURIComponent(datasetName)}/risk-areas`;

    debugLog.log(`Fetching risk areas for ${datasetName}`);

    try {
      const response = await api.get<{ risk_areas: RiskAreaInfo[] }>(endpoint);
      const duration = Date.now() - startTime;
      debugLog.log(`Fetched ${response.risk_areas?.length || 0} risk areas in ${duration}ms`);
      return response.risk_areas || [];
    } catch (error) {
      debugLog.error(`Failed to fetch risk areas for ${datasetName}:`, error);
      throw error;
    }
  }

  /**
   * Get statistics for a dataset
   */
  async getStatistics(datasetName: string): Promise<BenchmarkStatistics> {
    const startTime = Date.now();
    const endpoint = `${this.baseUrl}/${encodeURIComponent(datasetName)}/statistics`;

    debugLog.log(`Fetching statistics for ${datasetName}`);

    try {
      const response = await api.get<BenchmarkStatistics>(endpoint);
      const duration = Date.now() - startTime;
      debugLog.log(`Fetched statistics in ${duration}ms`);
      return response;
    } catch (error) {
      debugLog.error(`Failed to fetch statistics for ${datasetName}:`, error);
      throw error;
    }
  }

  /**
   * Generate a test batch for adversarial testing
   */
  async generateTestBatch(
    datasetName: string,
    config: TestBatchConfig
  ): Promise<TestBatchResponse> {
    const startTime = Date.now();
    const endpoint = `${this.baseUrl}/${encodeURIComponent(datasetName)}/test-batch`;

    debugLog.log(`Generating test batch from ${datasetName}`, config);

    try {
      const response = await apiClient.post<TestBatchResponse>(endpoint, config);
      const duration = Date.now() - startTime;
      debugLog.log(`Generated ${response.data.count} prompts in ${duration}ms`);
      return response.data;
    } catch (error) {
      debugLog.error(`Failed to generate test batch from ${datasetName}:`, error);
      throw error;
    }
  }

  /**
   * Export prompts for jailbreak testing
   */
  async exportForJailbreak(
    datasetName: string,
    config?: ExportConfig
  ): Promise<{ data: string; format: string; count: number }> {
    const startTime = Date.now();
    const endpoint = `${this.baseUrl}/${encodeURIComponent(datasetName)}/export`;

    debugLog.log(`Exporting prompts from ${datasetName}`, config);

    try {
      const response = await apiClient.post<{ data: string; format: string; count: number }>(
        endpoint,
        config || { format: "json" }
      );
      const duration = Date.now() - startTime;
      debugLog.log(`Exported ${response.data.count} prompts in ${duration}ms`);
      return response.data;
    } catch (error) {
      debugLog.error(`Failed to export prompts from ${datasetName}:`, error);
      throw error;
    }
  }
}

// Export singleton instance
export const benchmarkDatasetService = new BenchmarkDatasetService();
